### ISSUE TITLE
Support for arm64 Macs

### ISSUE THREAD

**Issue by KishanBagaria** [node-ffi-napi/node-ffi-napi#118]
We're building an Electron app for arm64 Macs and running into the following issue while rebuilding native dependencies (one of which depends on ffi-napi):

```sh
⠸ Building module: ffi-napi, Completed: 0gyp info find Python using Python version 3.9.1 found at "/usr/local/opt/python@3.9/bin/python3.9"
⠼ Building module: ffi-napi, Completed: 0gyp http GET https://www.electronjs.org/headers/v11.1.0/node-v11.1.0-headers.tar.gz
⠇ Building module: ffi-napi, Completed: 0gyp http 200 https://www.electronjs.org/headers/v11.1.0/node-v11.1.0-headers.tar.gz
⠼ Building module: ffi-napi, Completed: 0gyp http GET https://www.electronjs.org/headers/v11.1.0/SHASUMS256.txt
⠦ Building module: ffi-napi, Completed: 0gyp http 200 https://www.electronjs.org/headers/v11.1.0/SHASUMS256.txt
gyp info spawn /usr/local/opt/python@3.9/bin/python3.9
gyp info spawn args [
gyp info spawn args   '/Users/kishan/Downloads/foo/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/kishan/Downloads/foo/node_modules/ffi-napi/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/kishan/Downloads/foo/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/kishan/Downloads/foo/node_modules/ffi-napi/11.1.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/kishan/Downloads/foo/node_modules/ffi-napi/11.1.0',
gyp info spawn args   '-Dnode_gyp_dir=/Users/kishan/Downloads/foo/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/kishan/Downloads/foo/node_modules/ffi-napi/11.1.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/kishan/Downloads/foo/node_modules/ffi-napi',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
⠹ Building module: ffi-napi, Completed: 0gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  CC(target) Release/obj.target/nothing/../node-addon-api/src/nothing.o
⠸ Building module: ffi-napi, Completed: 0  LIBTOOL-STATIC Release/nothing.a
⠼ Building module: ffi-napi, Completed: 0warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: archive library: Release/nothing.a the table of contents is empty (no object file members in the library define global symbols)
  CC(target) Release/obj.target/ffi/deps/libffi/src/prep_cif.o
../deps/libffi/src/prep_cif.c:26:10: fatal error: 'ffi.h' file not found
#include <ffi.h>
         ^~~~~~~
⠴ Building module: ffi-napi, Completed: 01 error generated.
make: *** [Release/obj.target/ffi/deps/libffi/src/prep_cif.o] Error 1
✖ Rebuild Failed

An unhandled error occurred inside electron-rebuild
node-gyp failed to rebuild '/Users/kishan/Downloads/foo/node_modules/ffi-napi'.
Error: `make` failed with exit code: 2



Error: node-gyp failed to rebuild '/Users/kishan/Downloads/foo/node_modules/ffi-napi'.
Error: `make` failed with exit code: 2


    at ModuleRebuilder.rebuildNodeGypModule (/Users/kishan/Downloads/foo/node_modules/electron-rebuild/lib/src/module-rebuilder.js:193:19)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Rebuilder.rebuildModuleAt (/Users/kishan/Downloads/foo/node_modules/electron-rebuild/lib/src/rebuild.js:190:9)
    at async Rebuilder.rebuild (/Users/kishan/Downloads/foo/node_modules/electron-rebuild/lib/src/rebuild.js:152:17)
    at async /Users/kishan/Downloads/foo/node_modules/electron-rebuild/lib/src/cli.js:146:9
```

Repro steps:
```sh
mkdir foo
cd foo
yarn add ffi-napi electron-rebuild electron
electron-rebuild --force --arch arm64
```

**Comment by KishanBagaria** [comment]
Can I provide more info or help with this issue in any way? Pointers would be appreciated!

**Comment by KishanBagaria** [comment]
Should mention the repro steps above are for a regular x64 Mac. Happy to provide remote access to an arm64 Mac if needed for testing.

**Comment by addaleax** [comment]
@KishanBagaria Does the `yarn add` step succeed? Does it print anything, and/or does `node_modules/ffi-napi/build/Release` contain any `*.node` files?

**Comment by KishanBagaria** [comment]
Yep, the `yarn add` step succeeds. `electron-rebuild` fails with the above error log.

These are the only `*.node` files:
```sh
$ find node_modules/ffi-napi -name '*.node'
node_modules/ffi-napi/prebuilds/win32-ia32/node.napi.uv1.node
node_modules/ffi-napi/prebuilds/linux-arm64/node.napi.uv1.armv8.node
node_modules/ffi-napi/prebuilds/darwin-x64/node.napi.uv1.node
node_modules/ffi-napi/prebuilds/win32-x64/node.napi.uv1.node
node_modules/ffi-napi/prebuilds/linux-x64/node.napi.uv1.node
```

It appears to be looking for these files that are missing:
```
node_modules/ffi-napi/deps/libffi/config/mac/arm64/ffi.h
node_modules/ffi-napi/deps/libffi/config/mac/arm64/fficonfig.h
node_modules/ffi-napi/deps/libffi/config/mac/arm64/ffitarget.h
```

**Comment by KishanBagaria** [comment]
If I use `ffi.h` and `ffitarget.h` from `ffi-napi/deps/libffi/config/linux/arm64` and `fficonfig.h` from `ffi-napi/deps/libffi/config/mac/x64`, it rebuilds fine but crashes when running it on an arm64 Mac:

![image](https://user-images.githubusercontent.com/1093313/104629236-3f8ed680-56bf-11eb-8d2f-348fe84628a3.png)

**Comment by doug-a-brunner** [comment]
Not sure if related, but I'd been getting intermittent build errors in our Ubuntu Bionic CI environment with `ffi-napi@3.1.0`. It would always die with `No rule to make target 'Release/obj.target/nothing/../node-addon-api/src/nothing.o', needed by 'Release/obj.target/../node-addon-api/src/nothing.a'.  Stop.`. My guess is that it wasn't ensuring that the invocation of `gyp_main.py` finished before running `make`. Fortunately, the problem seems to have gone away after upgrading to `ffi-napi@4.0.1`.

**Comment by indutny** [comment]
@doug-a-brunner interesting. Unintended side-effects? 😂

**Comment by indutny** [comment]
This issue should be resolved in 4.0.1. Thank you so much for submitting in the first place!
